### PR TITLE
[Backend][FollowApi][3/n] Add unfollow and update follow api

### PR DESCRIPTION
**Pull request by garywangop**
1. Follow a user: 
正常follow: 
正常follow

follow已经follow过的user：
follow已follow的

follow不存在的user
follow不存在

2. Get all followers:
得到user全部的followers，按最新到最久的follow顺序排列
get all followers

查询一个不存在的用户：
查询不存在

3. Unfollow:
正常unfollow:
正常unfollow

unfollow一个没有follow的user：
unfollow没follow的user

### CODE REVIEW

File: http-service/src/main/java/com/laiteam/echowall/httpservice/api/FollowApi.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            throw new InvalidRequestException("Can't find user");
+        }
+        List<Profile> followerList = followService.getAllFollowers(userId, page, size);
+        return ResponseEntity.ok(FollowResponse.convertFollowers(userId, followerList));
+    }
+
+    @PostMapping(path = "/follow/{followId}")
+    public ResponseEntity<?> follow(@AuthenticationPrincipal User user, @PathVariable Long followId) throws Throwable {
+        Long userId = user.getId();
+        if (!profileService.getProfileById(followId).isPresent()) {
```
Comment (jingweihu) [inline http-service/src/main/java/com/laiteam/echowall/httpservice/api/FollowApi.java:38]: use userService to check the follower existance.

File: http-service/src/main/java/com/laiteam/echowall/httpservice/api/FollowApi.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                                         @RequestParam(defaultValue = "10", name = "size") Integer size) {
+        if (!profileService.findByUserId(userId).isPresent()) {
+            throw new InvalidRequestException("Can't find user");
+        }
+        List<Profile> followerList = followService.getAllFollowers(userId, page, size);
+        return ResponseEntity.ok(FollowResponse.convertFollowers(userId, followerList));
+    }
+
+    @PostMapping(path = "/follow/{followId}")
+    public ResponseEntity<?> follow(@AuthenticationPrincipal User user, @PathVariable Long followId) throws Throwable {
```
Comment (jingweihu) [inline http-service/src/main/java/com/laiteam/echowall/httpservice/api/FollowApi.java:36]: what's return type?

File: service/src/main/java/com/laiteam/echowall/service/imp/FollowServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public Optional<Follow> followUser(Long userId, Long followId) {
+        Follow follow = Follow.builder().userId(userId).followId(followId).build();
+        return Optional.of(followRepository.save(follow));
+    }
+
+    @Override
+    public List<Profile> getAllFollowers(Long userId, Integer page, Integer size) {
```
Comment (jingweihu) [inline service/src/main/java/com/laiteam/echowall/service/imp/FollowServiceImpl.java:35]: getAllFollowers(Long userId, PageRequest)?

File: service/src/main/java/com/laiteam/echowall/service/imp/FollowServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Override
+    @Transactional
+    public Optional<Profile> deleteFollower(Long userId, Long followId) {
+        followRepository.deleteFollowByFollowId(userId, followId);
+        return profileRepository.findById(followId);
+    }
+
+    @Override
+    public Optional<Follow> getFollower(Long userId, Long followId) {
```
Comment (jingweihu) [inline service/src/main/java/com/laiteam/echowall/service/imp/FollowServiceImpl.java:48]: isFollowed?

File: service/src/main/java/com/laiteam/echowall/service/FollowService.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import com.laiteam.echowall.dal.entity.Follow;
+import com.laiteam.echowall.dal.entity.Profile;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface FollowService {
+    Optional<Follow> followUser(Long userId, Long followId);
+
+    List<Profile> getAllFollowers(Long userId, Integer page, Integer size);
```
Comment (jingweihu) [inline service/src/main/java/com/laiteam/echowall/service/FollowService.java:12]: List of PagableInfo?

File: http-service/src/main/java/com/laiteam/echowall/httpservice/response/FollowResponse.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import com.laiteam.echowall.httpservice.utils.BeanUtils;
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Setter
+@Getter
+public class FollowResponse {
```
Comment (jingweihu) [inline http-service/src/main/java/com/laiteam/echowall/httpservice/response/FollowResponse.java:13]: I don't think you have to create a new FollowResponse using ProfileResponse should be enough

**Review by jingweihu — CHANGES_REQUESTED** [review]

**Comment by jingweihu** [comment]
pretty close. 
First, ProfileResponse should be enough for FollowResponse, you could add more field there 
Second, getAllFollows response should have page info, otherwise client won't know how many pages it has fetched.

File: http-service/src/main/java/com/laiteam/echowall/httpservice/api/FollowApi.java (hunk 1 above)
Comment (garywangop) [reply http-service/src/main/java/com/laiteam/echowall/httpservice/api/FollowApi.java:38]: Fixed

**Review by jingweihu — APPROVED** [review]

**Pull request merged**